### PR TITLE
Closes #2585 roclet: implement roclet to improve examples in documentation

### LINKED ISSUE #2585: General Issue: Implement roxygen roclet to add titles and descriptions to examples
### Background Information

A roxygen roclet defines how the roxygen tags in the function header are processed and how the man pages are generated. A roclet should be implemented such that a `@title` (and a `@description` tag) can be added for examples. Then each example gets a subsection with title and optionally a description. This helps navigating documentation of functions with many long examples like [derive_var_extreme_flag()](https://pharmaverse.github.io/admiral/reference/derive_var_extreme_flag.html#ref-examples).
See https://roxygen2.r-lib.org/articles/extending.html and https://roxygen2.r-lib.org/reference/roclet.html for some guidance on implementing roxygen roclets.

Benefits:
- More examples can be added without loosing readability.
- Finding and navigating the examples is improved.
- The reference pages can be used to train AI/LLM tools which should propose admiral function calls from plain English descriptions.

### Definition of Done

A roclet is implemented.

### PR DESCRIPTION
**Pull request by bundfussr**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF and a PNG version of it to the same folder. (The PNG version can be created by taking a screenshot of the PDF version.)
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by bms63** [comment]
image

image

**Comment by bundfussr** [comment]
@bms63 , @manciniedoardo , @ddsjoberg , @Fanny-Gautier , @StefanThoma , I have added an example section using markdown instead of the new tags in `R/derive_extreme_records.R` from line 259 onwards.

**Comment by StefanThoma** [comment]
> @bms63 , @manciniedoardo , @ddsjoberg , @Fanny-Gautier , @StefanThoma , I have added an example section using markdown instead of the new tags in `R/derive_extreme_records.R` from line 259 onwards.

Thanks for the update. 
I kind of prefer the markdown version, because I'm more familiar with it, but also because it seems more consistent with the R package as it's used throughout. 
However, I don't quite know its limitations. Is there anything that roclet can do that cannot be achieved easily with markdown? Are there differences in execution times?

Oh and roclet reminds me of:

**Comment by bundfussr** [comment]
> However, I don't quite know its limitations. Is there anything that roclet can do that cannot be achieved easily with markdown? Are there differences in execution times?

At the moment there is no difference in the output or the execution times.

I think the structure is clearer using the new tags but it may be a bit overkill to introduce new tags just for this.

However, the tags give us opportunities for further improvements. For example:
- The examples could be written to a separate file which can be used to train LLMs. The structure of the file could be tailored to whatever works best to train the LLM.
- We could enable linking to examples. Then in the argument description we could add links to examples that illustrate the use of the argument.
- We could enable adding tags (not roxygen tags) to examples and then generate a separate page presenting examples by tag. For instance we could add a "dtype" tag to all examples deriving `DTYPE`:
    ```
    @caption Deriving records for `DTYPE = "LOV"`
    @tags dtype
    ...
    ```
    And then create a separate page showing all examples with the "dtype" tag.

**Comment by bms63** [comment]
Hey all I have some thoughts but want to review PR more, but it won't be until the end of the week.

**Comment by bundfussr** [comment]
I've moved the roclet to admiraldev (see https://github.com/pharmaverse/admiraldev/pull/485).

**Comment by bundfussr** [comment]
I have updated the examples for `derive_extreme_records()` as an suggestion how we could structure our examples.

There are two parts: 
- The first part are examples which demonstrate the functionality of the function. The first example shows the default behavior. Then examples for each optional argument follows. The data in these examples is as concise as possible and may not be ADaM compliant, e.g., `STUDYID` is not included or `AVISITN` is present without `AVISIT`.
These examples should answer questions like "what does this function?" or "how works this argument?"
- The second part are examples which demonstrate which ADaM variables or parameters could be derived with the function. The code could be copied to an ADaM script and be used without or with small modifications.
Each example includes a "specs-like" description.
These examples should answer questions like "how can this variable/argument be derived".

Please have a look.

**Comment by manciniedoardo** [comment]
@pharmaverse/admiral 

@bms63 and I met today to discuss this topic in detail and propose a strategy moving forward. This below is what we've settled on - please read through and let us know any thoughts.
______________________________________________________

- The debate is regarding our examples in the reference pages and whether to implement the roclet approach as showcased by @bundfussr in this PR, or just including standard markdown for our examples instead.

- On the whole, we prefer the roclet approach, because:
  - It is somewhat cleaner 
  - It is easier to "police" which tags are being used and how
  - It could have a lot of future potential (e.g. with training LLMs)
 
- At the same time we recognise that selecting this approach carries some challenges and risks:
  - It makes the documentation process more complicated, especially if we come back to it after a few years of less activity on the package.
  - As far as we understand from @ddsjoberg 's comments at last week's meeting (@ddsjoberg please correct if we're mistaken),  the roclet approach (but actually even the RMarkdown approach) carries CRAN-related risks, namely that in theory CRAN could turn around in 6 months/1 year/2 years and request that we provide standard example for our functions. This would mean that we'd have to go back and downgrade all the examples back to how they are now, Packages like {gt} choose to run this risk and live dangerously, but we'd like to have a fallback option if this were to happen.

As such we propose the following:
- Select the roclet approach, on the proviso that a "revert" script can be created first, which takes in our "rocletised" examples and spits out examples that CRAN can accept. @bundfussr @ddsjoberg do you think that's possible?
- Implement the new examples for one function (maybe by merging this PR  for `derive_extreme_records()` 
- Make a tier list of first, second and third tier functions whose examples we'll improve. For instance, the first tier may include `derive_vars_joined()`, `derive_param_computed()`, etc), the second tier may have functions like `derive_vars_crit_flag()` and the third tier could contain the lowest priority functions, e.g. the computation functions. I can take on proposing this tier list.
- Chip away at the first and second tier function examples throughout this release period and see where we are at in June.

Sorry for the wall of text!
image

File: R/derive_extreme_records.R
Comment (bms63) [inline R/derive_extreme_records.R:1]: Let's all discuss on this!!  @pharmaverse/admiral

**Comment by bms63** [comment]
I have in my notes we are supposed to review one of the extreme functions.  I haven't done that yet.  If you are fine and others on @pharmaverse/admiral are happy with it then lets proceed

Also - we need our doomsday code in case CRAN fights us about the change.  Hoping it would be something simple to revert the changes back to original roxygen format.  I feel that this minimizes disruption to users and prevents later dev time spent on this revert.

**Comment by bundfussr** [comment]
@manciniedoardo , @bms63 , I thought we would discuss the way how we want to structure our examples (regarding the content not the tags) in the developer meeting on Monday (see https://github.com/pharmaverse/admiral/pull/2660#issuecomment-2672188314 for my proposal).

 For merging this MR we need to merge https://github.com/pharmaverse/admiraldev/pull/485 first. The main functionality (in admiraldev) is implemented but two things are missing:
- The vignettes are not updated yet.
- I would like to improve the handling of errors and warnings. By default, an example which issues a warning or error should cause the building of the documentation to fail. If the warning or error is expected it should be specified in the `@code` tag.

Regarding the "doomsday code" I'm not sure what exactly we should do. If CRAN insists on having `\examples{}` calls in the man files we have several options:
- Keep our tags (`@caption`, `@info`, `@code`) but write everything to an `\examples{}` call (`@caption` and `@info` as comments).
- Update all headers (by a script) and replace our tags.
- Use `\examples{}` or `@examples` for the first example and keep our tags for all other examples. This could be implemented within the roclet or by a script which changes the headers.
- Add a dummy `\examples{}` or `@examples` entry with a comment like `# For examples see above`. Also this could be implemented via roclet or via changing the headers.

Should we implement anything now? How likely is it that CRAN insists on `\examples{}` calls?
I assume they have this requirement because a documentation without examples is hard to understand. However, we don't have `\examples{}` calls because we don't have examples but because we want more and better examples.
Therefore, I would tend to not implementing something now.

**Comment by bms63** [comment]
> > Should we implement anything now? How likely is it that CRAN insists on \examples{} calls?
> 
> I think it's very unlikely. other packages run this risk too. I like this idea:
> 
> > Use \examples{} or @examples for the first example and keep our tags for all other examples. This could be implemented within the roclet or by a script which changes the headers.
> 
> Because we could easily do this reactively should it come to it, but wouldn't need to do anything now. Alternatively the dummy @examples would work too.
> 
> Let's prioritise [this PR](https://github.com/pharmaverse/admiraldev/pull/485) and discuss any issues/oustanding work on our call Monday, with a look to then try merge this one after.

i only know of gt that does this rouge behavior.  what are the other packages?

**Comment by manciniedoardo** [comment]
> i only know of gt that does this rouge behavior. what are the other packages?

@ddsjoberg did you mention some others?

File: NEWS.md [hunk 1]
```diff
@@ -2,10 +2,20 @@
...
 ## New Features
 
+- The documentation was enhanced: (#2585)
+    - The default value of an argument is now displayed in the argument description.
+    - For complex functions each example has now a title, which is also shown in
+    the TOC, and a description. This enabled adding more examples without losing
+    readability. See `derive_extreme_records()` for an example.
+    - The output of the structured examples used for complex functions is
+    displayed in the help pages in RStudio.
+
```
Comment (manciniedoardo) [inline NEWS.md:12]: I think it might be a good idea to also then include under "Updates to documentation" a line for each of the functions that has undergone an examples revamp. So, in this PR, we'd add derive_extreme_records, but if we get to more functions by the release, we can list them there.

File: R/derive_extreme_records.R [hunk 2]
```diff
@@ -11,9 +11,12 @@
 #'
```
Comment (manciniedoardo) [inline R/derive_extreme_records.R:11]: Should we say here that `dataset` can be left `NULL`, in which case a dataset will be created? It initially confused me that there was no `dataset` in some of the examples, for instance the section "Flagging existence of source records (exist_flag, true_value, false_value)"

File: R/derive_extreme_records.R [hunk 3]
```diff
@@ -62,7 +67,7 @@
 #'   For all other by groups `exist_flag` is set to the value specified by the
 #'   `false_value` argument.
 #'
-#'   *Permitted Values:* Variable name
+#' @permitted Variable name
 #'
 #' @param true_value True value
 #'
```
Comment (manciniedoardo) [inline R/derive_extreme_records.R:73]: Mention in the details below also that if `dataset` is not passed, a new dataset is created.

**Review by manciniedoardo — CHANGES_REQUESTED** [review]
Looks awesome! I've left some comments regarding how to best organise the news section and also one about the new examples section for `derive_extreme_records()`

File: NEWS.md (hunk 1 above)
Comment (bundfussr) [reply NEWS.md:12]: Yes, item added.

File: R/derive_extreme_records.R (hunk 2 above)
Comment (bundfussr) [reply R/derive_extreme_records.R:11]: Done

File: R/derive_extreme_records.R (hunk 3 above)
Comment (bundfussr) [reply R/derive_extreme_records.R:84]: Done

**Review by manciniedoardo — APPROVED** [review]
Looking good!

File: NEWS.md [hunk 4]
```diff
@@ -22,9 +32,12 @@
 
 ## Documentation
 
+- The examples section of the following functions was enhanced:
+`derive_extreme_records()` (#2585)
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply NEWS.md]: Updated
